pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/CursorIntegrationTest.java

```diff
@@ -0,0 +1,432 @@
+
+    // Get a list of responses using cursors.
+    CursorResponse pinotPagingResponse;
+    pinotPagingResponse = JsonUtils.jsonNodeToObject(ClusterTest.postQuery(pinotQuery,
+        getBrokerQueryApiUrl(queryResourceUrl) + getCursorQueryProperties(_resultSize), headers,
+        getExtraQueryProperties()), CursorResponseNative.class);
+    if (!pinotPagingResponse.getExceptions().isEmpty()) {
+      throw new RuntimeException("Got Exceptions from Query Response: " + pinotPagingResponse.getExceptions().get(0));
+    }
+    List<CursorResponse> resultPages = getAllResultPages(queryResourceUrl, headers, pinotPagingResponse, _resultSize);
+
+    int brokerPagingResponseSize = 0;
+    for (CursorResponse response : resultPages) {
+      brokerPagingResponseSize += response.getNumRows();
+    }
+
+    // Compare the number of rows.
+    if (brokerResponseSize != brokerPagingResponseSize) {
+      throw new RuntimeException(
+          "Pinot # of rows from paging API " + brokerPagingResponseSize + " doesn't match # of rows from default API "
+              + brokerResponseSize);
+    }
+  }
+
+  private List<CursorResponse> getAllResultPages(String queryResourceUrl, Map<String, String> headers,
+      CursorResponse firstResponse, int numRows)
+      throws Exception {
+    numRows = numRows == 0 ? CommonConstants.CursorConfigs.DEFAULT_CURSOR_FETCH_ROWS : numRows;
+
+    List<CursorResponse> resultPages = new ArrayList<>();
+    resultPages.add(firstResponse);
+    int totalRows = firstResponse.getNumRowsResultSet();
+
+    int offset = firstResponse.getNumRows();
+    while (offset < totalRows) {
+      CursorResponse response = JsonUtils.stringToObject(ClusterTest.sendGetRequest(
+          getBrokerResponseApiUrl(queryResourceUrl, firstResponse.getRequestId()) + getCursorOffset(offset, numRows),
+          headers), CursorResponseNative.class);
+      resultPages.add(response);
+      offset += response.getNumRows();
+    }
+    return resultPages;
+  }
+
+  protected Object[][] getPageSizesAndQueryEngine() {
+    return new Object[][]{
+        {false, 2}, {false, 3}, {false, 10}, {false, 0}, //0 trigger default behaviour
+        {true, 2}, {true, 3}, {true, 10}, {true, 0} //0 trigger default behaviour
+    };
+  }
+
+  @DataProvider(name = "pageSizeAndQueryEngineProvider")
+  public Object[][] pageSizeAndQueryEngineProvider() {
+    return getPageSizesAndQueryEngine();
+  }
+
+  // Test hard coded queries with SSE/MSE AND different cursor response sizes.
+  @Test(dataProvider = "pageSizeAndQueryEngineProvider")
+  public void testHardcodedQueries(boolean useMultiStageEngine, int pageSize)
+      throws Exception {
+    _resultSize = pageSize;
+    setUseMultiStageQueryEngine(useMultiStageEngine);
+    super.testHardcodedQueries();
+  }
+
+  @DataProvider(name = "chooseQueryEngine")
+  public Object[][] chooseQueryEngine() {
+    return new Object[][] {
+        {false}, {true}
+    };
+  }
```

Comment:
This already exists (`useBothQueryEngines`).

Reply:
Removed